packages/base/src/lazy-tsp-client.ts

```diff
@@ -0,0 +1,33 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
+
+import { TspClient } from 'tsp-typescript-client';
+
+export type LazyTspClient = {
+    [K in keyof TspClient]: TspClient[K] extends (...args: infer A) => (infer R | Promise<infer R>)
+        ? (...args: A) => Promise<R>
+        : never // discard field
+};
+
+export const LazyTspClientFactory = Symbol('LazyTspClientFactory');
+export type LazyTspClientFactory = (url: Promise<string>) => LazyTspClient;
+
+export function LazyTspClientFactoryImpl(url: Promise<string>): LazyTspClient {
+    // Most(all) methods from the `TspClient` are asynchronous.
+    // The `LazyTspClient` will just delay each call to its methods by
+    // first awaiting for the asynchronous `baseUrl` resolution which
+    // is used to then asynchronously get a valid `TspClient`.
+    const tspClientPromise = url.then(baseUrl => new TspClient(baseUrl));
+    // eslint-disable-next-line no-null/no-null
+    return new Proxy(Object.create(null), {
+        get(target, property, _receiver) {
+            let method = target[property];
+            if (!method) {
+                target[property] = method = async (...args: any[]) => {
```

Comment:
I had to learn how the proxy works and I think I understand how it works. Basically, with your code each method call is intercepted here and forwarded to the corresponding TspClient object's method while making sure that baseUrl is set properly first. Thanks for adding the comment above that helps to understand.

One thing though, the code will fail to compile the day the TspClient has one synchronous method. Do you see any issues with that? Or do we just make sure that all new methods in the TspClient is async until the LazyTspClient is not needed anymore? Please let me know what you think.

Reply:
Yes the idea was to delay the method calls on the `LazyTspClient` until the time it is possible to instantiate a `TspClient`, which we do once the promise to an URL resolves.

On the other hand, with the way I wrote things here compilation won't fail even if you add synchronous methods... It will most likely fail at runtime. But ultimately what this means is that this code cannot handle synchronous calls, not as it is right now.

Reply:
I type a lot don't I.

> Or do we just make sure that all new methods in the TspClient is async until the LazyTspClient is not needed anymore?

It sucks but that's one way to go. The best would be to get rid of the `LazyTspClient` before the time you need to add synchronous methods to the `TspClient` ;)

Otherwise on a per-case basis it might be possible to try and intercept the calls to the synchronous APIs and try to provide a somewhat safe value/function, but that starts to become smelly.

Reply:
I tried it with yarn link "tsp-typescript-client" and a synchronous method, and it fail at compile time. That's better than runtime.

Reply:
Oh that's surprising, cool.

Reply:
Found it. It caused an error when forcing the type `as TspClient`. I moved types around so that the error happens in the factory instead of at call sites.